python/TestHarness/schedulers/Scheduler.py

```diff
@@ -0,0 +1,491 @@
+            if self.workers_in_use + tester.getProcs(self.options) <= self.job_slots:
+                return True
+
+            # Check for insufficient slots -soft limit
+            # TODO: Create a unit test for this case
+            elif tester.getProcs(self.options) > self.job_slots and self.soft_limit:
+                tester.specs.addParam('caveats', ['OVERSIZED'], "")
+                return True
+
+            # Check for insufficient slots -hard limit (skip this job)
+            # TODO: Create a unit test for this case
+            elif tester.getProcs(self.options) > self.job_slots and not self.soft_limit:
+                tester.setStatus('insufficient slots', tester.bucket_skip)
+
+    # Get next available set of jobs
+    def getNextJobGroup(self, tester_data):
+        tester = tester_data.getTester()
+        tester_dag = tester_data.getDAG()
+        next_group = []
+
+
+        # Handle dependencies that will fail if this test did not successfully finish but only if we care
+        # about prereqs
+        if not tester.didPass():
+            self.deleteDownstreamTests(self.scheduled_tester_data[tester.getTestName()])
+
+        with self.thread_lock:
+            # Get the next list of runnable nodes (jobs)
+            tester_dag.delete_node(tester.getTestName())
+            next_jobs = tester_dag.ind_nodes()
+            # Loop through and determine if we need to assign a thread to this new job
+
+            for next_job in next_jobs:
+                tmp_tester = self.scheduled_tester_data[next_job].getTester()
+                if next_job not in self.active_jobs and tmp_tester.isInitialized():
+                    tmp_tester.setStatus('QUEUED', tmp_tester.bucket_pending)
+                    next_group.append(next_job)
+
+        return next_group
+
+    # put a job in the runner queue and assign a runner thread to do some work
+    def assignRunner(self, job):
+        self.runner_queue.put(job)
+        return self.runner_pool.apply_async(self.jobRunner, (self.runner_queue,))
+
+    # put a job in the status queue and assign a status thread to do some work
+    def assignStatus(self, job):
+        self.status_queue.put(job)
+        return self.status_pool.apply_async(self.statusRunner, (self.status_queue,))
+
+    # Status processing (hand a tester's status back to the TestHarness to have some status printed to the screen)
+    def statusRunner(self, queue):
+        job = queue.get()
+        self.harness.handleTestStatus(self.scheduled_tester_data[job])
+
+    # Runner processing (call derived run method using a thread)
+    def jobRunner(self, queue):
+        job = queue.get()
+        tester_dag = self.scheduled_tester_data[job].getDAG()
+        tester = self.scheduled_tester_data[job].getTester()
+
+        # Check if we have available slots
+        if self.checkAvailableSlots(tester):
+            can_run = True
+            with self.thread_lock:
```

Comment:
NOPE! You created a race condition here that's hard to see because of the subroutine you called. You have to be very careful with locks and subroutines. In `checkAvailableSlots()`, you lock and inspect at the shared resource `workers_in_use` to figure out whether you can run. However, you then let go of the lock as you exit the routine after you've answered the question. Another thread can sneak in there and also think it's OK to start. You have to keep the inspection and writing of the resource together. This is paramount.